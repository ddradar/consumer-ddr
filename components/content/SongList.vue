<template>
  <OTable :data="songs!" striped narrowed :loading="isLoading">
    <OTableColumn v-slot="props" field="name" label="Name">
      <NuxtLink :to="`/songs/${props.row.slug}/`">
        {{ props.row.name }}
      </NuxtLink>
    </OTableColumn>
    <OTableColumn v-slot="props" field="artist" label="Artist">
      <span>{{ props.row.artist }}</span>
    </OTableColumn>
    <OTableColumn v-slot="props" field="bpm" label="BPM">
      {{ props.row.bpm }}
    </OTableColumn>
    <OTableColumn
      v-for="playStyle in playStyles"
      v-slot="props"
      :key="playStyle"
      :label="playStyle"
    >
      <div class="tags has-addons">
        <span
          v-for="chart in props.row.charts[playStyle]"
          :key="chart.difficulty"
          class="tag"
          :class="chart.color"
          :title="chart.name"
        >
          {{ chart.level }}
        </span>
      </div>
    </OTableColumn>

    <template #empty>
      <section class="section">
        <div class="content has-text-grey has-text-centered">
          <p>Nothing here.</p>
        </div>
      </section>
    </template>
  </OTable>
</template>

<script lang="ts" setup>
import type {
  Chart,
  PlayStyle,
  SoftwareParsedContent,
  SongParsedContent
} from '~~/src/content'
import { normalizeDifficulty } from '~~/src/song'

type ChartInfo = Pick<Chart, 'level' | 'difficulty'> & {
  color: string
  name: string
}
interface Props {
  series: string
}

const prop = defineProps<Props>()

const { data: software } = await useAsyncData(`/software/${prop.series}`, () =>
  queryContent<SoftwareParsedContent>(prop.series)
    .where({ _type: 'markdown' })
    .findOne()
)

const { data: songs, pending: isLoading } = await useAsyncData(
  `/software/${prop.series}/songs`,
  () =>
    queryContent<SongParsedContent>(prop.series)
      .where({ _type: 'json' })
      .without('series')
      .find(),
  {
    transform: (songs: SongParsedContent[]) =>
      songs.map((s) => ({
        slug: s.slug,
        name: s.name,
        artist: s.artist,
        bpm: s.bpm,
        charts: s.charts.reduce((p, c) => {
          if (!p[c.playStyle]) {
            p[c.playStyle] = []
          }
          const name =
            software.value?.difficultyNames[c.difficulty] ?? 'Unknown'
          p[c.playStyle]!.push({
            level: c.level,
            difficulty: c.difficulty,
            name,
            color: `is-${normalizeDifficulty(name).toLowerCase()}`
          })
          return p
        }, {} as Partial<Record<PlayStyle, ChartInfo[]>>)
      }))
  }
)

const playStyles = computed(() => [
  ...new Set(songs.value?.flatMap((s) => Object.keys(s.charts)))
])
</script>
