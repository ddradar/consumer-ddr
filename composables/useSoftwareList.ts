import type { Ref } from 'vue'

import type { Software } from '~~/src/content'

const keys = [
  'slug',
  'title',
  'color',
  'platform',
  'region',
  'launched'
] as const
export type SoftwareListData = Pick<Software, typeof keys[number]>

export default async function () {
  const { data } = await useAsyncData('/software', () =>
    queryContent<Software>('series')
      .sort({ launched: 1 })
      .only([...keys])
      .find()
  )

  return { softwareList: data as Ref<SoftwareListData[]> }
}
