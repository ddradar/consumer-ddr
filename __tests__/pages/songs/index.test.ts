import { RouterLinkStub } from '@vue/test-utils'
import { describe, expect, test, vi } from 'vitest'
import { ref } from 'vue'

import useSoftwareList from '~~/composables/useSoftwareList'
import paranoiaFirst from '~~/content/1st-jp/paranoia.json'
import am3p from '~~/content/2nd/am-3p.json'
import paranoiaSecond from '~~/content/2nd/paranoia.json'
import SongList from '~~/pages/songs/index.vue'

import { mockSoftwareList, mountAsync, plugins } from '../../test-utils'

vi.mock('~~/composables/useSoftwareList')

const rawSongData = [am3p, paranoiaFirst, paranoiaSecond].map((d) => ({
  slug: d.slug,
  series: d.series,
  name: d.name,
  artist: d.artist,
  bpm: d.bpm
}))

describe('pages/songs/index.vue', () => {
  const songs = [
    {
      slug: am3p.slug,
      name: am3p.name,
      artist: am3p.artist,
      bpm: am3p.bpm,
      seriesList: [mockSoftwareList[1]]
    },
    {
      slug: paranoiaFirst.slug,
      name: paranoiaFirst.name,
      artist: paranoiaFirst.artist,
      bpm: paranoiaFirst.bpm,
      seriesList: mockSoftwareList
    }
  ]
  const global = { plugins, stubs: { NuxtLink: RouterLinkStub } }

  describe('snapshot test', () => {
    test('renders loading state', async () => {
      // Arrange
      vi.mocked(useSoftwareList).mockResolvedValue({
        softwareList: ref(mockSoftwareList),
        isLoading: ref(false)
      })
      vi.mocked(useAsyncData).mockResolvedValue({
        data: ref([]),
        pending: ref(true)
      } as any)

      // Act
      const wrapper = await mountAsync(SongList, { global })

      // Assert
      expect(wrapper.element).toMatchSnapshot()
    })

    test('renders song info', async () => {
      // Arrange
      vi.mocked(useSoftwareList).mockResolvedValue({
        softwareList: ref(mockSoftwareList),
        isLoading: ref(false)
      })
      vi.mocked(useAsyncData).mockResolvedValue({
        data: ref(songs),
        pending: ref(false)
      } as any)

      // Act
      const wrapper = await mountAsync(SongList, { global })

      // Assert
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  describe('useAsyncData()', () => {
    test('calls with transform option', async () => {
      // Arrange
      vi.mocked(useSoftwareList).mockResolvedValue({
        softwareList: ref(mockSoftwareList),
        isLoading: ref(false)
      })
      vi.mocked(useAsyncData).mockReset()
      vi.mocked(useAsyncData).mockResolvedValue({
        data: ref(songs),
        pending: ref(false)
      } as any)

      // Act
      await mountAsync(SongList, { global })
      const transform = vi.mocked(useAsyncData).mock.calls[0][2]?.transform!

      // Assert
      expect(transform(rawSongData)).toStrictEqual(songs)
    })
  })
})
