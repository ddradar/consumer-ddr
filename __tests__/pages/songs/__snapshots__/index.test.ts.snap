// Vitest Snapshot v1

exports[`pages/songs/index.vue > snapshot test > renders loading state 1`] = `
<section
  class="section"
>
  <h1
    class="title"
  >
    Song List
  </h1>
  <div
    class="b-table"
  >
    <div
      style="display: none;"
    >
      
      <span
        data-id="1"
      >
        Name
      </span>
      <span
        data-id="2"
      >
        Artist
      </span>
      <span
        data-id="3"
      >
        BPM
      </span>
      <span
        data-id="4"
      >
        Series
      </span>
      
    </div>
    <!--v-if-->
    <!--v-if-->
    <div
      class="table-wrapper"
    >
      <table
        class="table is-striped is-narrow is-empty"
        tabindex="false"
      >
        <!--v-if-->
        <thead>
          <tr>
            <!--v-if-->
            <!--v-if-->
            
            <th
              class=""
              draggable="false"
            >
              <span>
                Name 
                <span
                  class="th-wrap sort-icon"
                  style="display: none;"
                >
                  <span
                    class="icon is-small"
                  >
                    <i
                      class="mdi mdi-arrow-up"
                    />
                  </span>
                </span>
              </span>
            </th>
            <th
              class=""
              draggable="false"
            >
              <span>
                Artist 
                <span
                  class="th-wrap sort-icon"
                  style="display: none;"
                >
                  <span
                    class="icon is-small"
                  >
                    <i
                      class="mdi mdi-arrow-up"
                    />
                  </span>
                </span>
              </span>
            </th>
            <th
              class=""
              draggable="false"
            >
              <span>
                BPM 
                <span
                  class="th-wrap sort-icon"
                  style="display: none;"
                >
                  <span
                    class="icon is-small"
                  >
                    <i
                      class="mdi mdi-arrow-up"
                    />
                  </span>
                </span>
              </span>
            </th>
            <th
              class=""
              draggable="false"
            >
              <span>
                Series 
                <span
                  class="th-wrap sort-icon"
                  style="display: none;"
                >
                  <span
                    class="icon is-small"
                  >
                    <i
                      class="mdi mdi-arrow-up"
                    />
                  </span>
                </span>
              </span>
            </th>
            
            <!--v-if-->
          </tr>
          <tr>
            <!--v-if-->
            <!--v-if-->
            
            <th
              class=""
            >
              
              <div
                class="control"
              >
                <input
                  autocomplete="off"
                  class="input"
                  on="function () { [native code] }"
                  type="text"
                />
                <!--v-if-->
                <!--v-if-->
                <!--v-if-->
              </div>
              
            </th>
            <th
              class=""
            >
              
              <div
                class="control"
              >
                <input
                  autocomplete="off"
                  class="input"
                  on="function () { [native code] }"
                  type="text"
                />
                <!--v-if-->
                <!--v-if-->
                <!--v-if-->
              </div>
              
            </th>
            <th
              class=""
            >
              <!--v-if-->
            </th>
            <th
              class=""
            >
              <!--v-if-->
            </th>
            
            <!--v-if-->
          </tr>
          <!--v-if-->
        </thead>
        <tbody>
          
          
          <tr>
            <td
              colspan="4"
            >
              
              <section
                class="section"
              >
                <div
                  class="content has-text-grey has-text-centered"
                >
                  <p>
                    Nothing here.
                  </p>
                </div>
              </section>
              
            </td>
          </tr>
        </tbody>
        <!--v-if-->
      </table>
      
      <transition-stub>
        <div
          class="o-load loading"
        >
          <div
            class="o-load__overlay loading-overlay"
          />
          
          <span
            class="icon is-spin is-medium o-load__icon icon"
          >
            <i
              class="mdi mdi-loading mdi-36px"
            />
          </span>
          
        </div>
      </transition-stub>
      
    </div>
    
    <div
      class="level"
    >
      <div>
        
        
      </div>
      <div>
        <div>
          <nav
            class="pagination"
          >
            <a
              aria-current="false"
              class="pagination-link pagination-previous is-disabled pagination-previous is-disabled"
              href="#"
              role="button"
            >
              
              <span
                aria-hidden="true"
                class="icon"
              >
                <i
                  class="mdi mdi-chevron-left mdi-24px"
                />
              </span>
              
            </a>
            <a
              aria-current="false"
              class="pagination-link pagination-next is-disabled pagination-next is-disabled"
              href="#"
              role="button"
            >
              
              <span
                aria-hidden="true"
                class="icon"
              >
                <i
                  class="mdi mdi-chevron-right mdi-24px"
                />
              </span>
              
            </a>
            <ul
              class="pagination-list"
            >
              <!--First-->
              <!--v-if-->
              <!--v-if-->
              <!--Pages-->
              
              
              <!--Last-->
              <!--v-if-->
              <!--v-if-->
            </ul>
          </nav>
        </div>
      </div>
    </div>
    
  </div>
</section>
`;

exports[`pages/songs/index.vue > snapshot test > renders song info 1`] = `
<section
  class="section"
>
  <h1
    class="title"
  >
    Song List
  </h1>
  <div
    class="b-table"
  >
    <div
      style="display: none;"
    >
      
      <span
        data-id="1"
      >
        Name
      </span>
      <span
        data-id="2"
      >
        Artist
      </span>
      <span
        data-id="3"
      >
        BPM
      </span>
      <span
        data-id="4"
      >
        Series
      </span>
      
    </div>
    <!--v-if-->
    <!--v-if-->
    <div
      class="table-wrapper"
    >
      <table
        class="table is-striped is-narrow"
        tabindex="false"
      >
        <!--v-if-->
        <thead>
          <tr>
            <!--v-if-->
            <!--v-if-->
            
            <th
              class=""
              draggable="false"
            >
              <span>
                Name 
                <span
                  class="th-wrap sort-icon"
                  style="display: none;"
                >
                  <span
                    class="icon is-small"
                  >
                    <i
                      class="mdi mdi-arrow-up"
                    />
                  </span>
                </span>
              </span>
            </th>
            <th
              class=""
              draggable="false"
            >
              <span>
                Artist 
                <span
                  class="th-wrap sort-icon"
                  style="display: none;"
                >
                  <span
                    class="icon is-small"
                  >
                    <i
                      class="mdi mdi-arrow-up"
                    />
                  </span>
                </span>
              </span>
            </th>
            <th
              class=""
              draggable="false"
            >
              <span>
                BPM 
                <span
                  class="th-wrap sort-icon"
                  style="display: none;"
                >
                  <span
                    class="icon is-small"
                  >
                    <i
                      class="mdi mdi-arrow-up"
                    />
                  </span>
                </span>
              </span>
            </th>
            <th
              class=""
              draggable="false"
            >
              <span>
                Series 
                <span
                  class="th-wrap sort-icon"
                  style="display: none;"
                >
                  <span
                    class="icon is-small"
                  >
                    <i
                      class="mdi mdi-arrow-up"
                    />
                  </span>
                </span>
              </span>
            </th>
            
            <!--v-if-->
          </tr>
          <tr>
            <!--v-if-->
            <!--v-if-->
            
            <th
              class=""
            >
              
              <div
                class="control"
              >
                <input
                  autocomplete="off"
                  class="input"
                  on="function () { [native code] }"
                  type="text"
                />
                <!--v-if-->
                <!--v-if-->
                <!--v-if-->
              </div>
              
            </th>
            <th
              class=""
            >
              
              <div
                class="control"
              >
                <input
                  autocomplete="off"
                  class="input"
                  on="function () { [native code] }"
                  type="text"
                />
                <!--v-if-->
                <!--v-if-->
                <!--v-if-->
              </div>
              
            </th>
            <th
              class=""
            >
              <!--v-if-->
            </th>
            <th
              class=""
            >
              <!--v-if-->
            </th>
            
            <!--v-if-->
          </tr>
          <!--v-if-->
        </thead>
        <tbody>
          
          
          <tr
            class=""
            draggable="false"
          >
            <!--v-if-->
            <!--v-if-->
            
            <td
              class=""
              data-label="Name"
              scoped=""
            >
              <a
                class="is-size-6-mobile"
              >
                AM-3P
              </a>
            </td>
            <td
              class=""
              data-label="Artist"
              scoped=""
            >
              <span
                class="is-size-6-mobile"
              >
                KTz
              </span>
            </td>
            <td
              class=""
              data-label="BPM"
              scoped=""
            >
              130
            </td>
            <td
              class=""
              data-label="Series"
              scoped=""
            >
              <div
                class="buttons"
              >
                
                <!--[object Object]-->
                
              </div>
            </td>
            
            <!--v-if-->
          </tr>
          <transition-stub>
            <!--v-if-->
          </transition-stub>
          <!--v-if-->
          
          
          <tr
            class=""
            draggable="false"
          >
            <!--v-if-->
            <!--v-if-->
            
            <td
              class=""
              data-label="Name"
              scoped=""
            >
              <a
                class="is-size-6-mobile"
              >
                PARANOiA
              </a>
            </td>
            <td
              class=""
              data-label="Artist"
              scoped=""
            >
              <span
                class="is-size-6-mobile"
              >
                180
              </span>
            </td>
            <td
              class=""
              data-label="BPM"
              scoped=""
            >
              180
            </td>
            <td
              class=""
              data-label="Series"
              scoped=""
            >
              <div
                class="buttons"
              >
                
                <!--[object Object]-->
                <!--[object Object]-->
                
              </div>
            </td>
            
            <!--v-if-->
          </tr>
          <transition-stub>
            <!--v-if-->
          </transition-stub>
          <!--v-if-->
          
          
          <!--v-if-->
        </tbody>
        <!--v-if-->
      </table>
      
      <transition-stub>
        <!--v-if-->
      </transition-stub>
      
    </div>
    
    <div
      class="level"
    >
      <div>
        
        
      </div>
      <div>
        <div>
          <nav
            class="pagination"
          >
            <a
              aria-current="false"
              class="pagination-link pagination-previous is-disabled pagination-previous is-disabled"
              href="#"
              role="button"
            >
              
              <span
                aria-hidden="true"
                class="icon"
              >
                <i
                  class="mdi mdi-chevron-left mdi-24px"
                />
              </span>
              
            </a>
            <a
              aria-current="false"
              class="pagination-link pagination-next is-disabled pagination-next is-disabled"
              href="#"
              role="button"
            >
              
              <span
                aria-hidden="true"
                class="icon"
              >
                <i
                  class="mdi mdi-chevron-right mdi-24px"
                />
              </span>
              
            </a>
            <ul
              class="pagination-list"
            >
              <!--First-->
              <!--v-if-->
              <!--v-if-->
              <!--Pages-->
              
              <li>
                <a
                  aria-current="true"
                  class="pagination-link is-current"
                  href="#"
                  role="button"
                >
                  
                  1
                  
                </a>
              </li>
              
              <!--Last-->
              <!--v-if-->
              <!--v-if-->
            </ul>
          </nav>
        </div>
      </div>
    </div>
    
  </div>
</section>
`;
